### PR TITLE
SGD learning rate tuning + pytorch implementation, simulations

### PR DESCRIPTION
**Pull request by jjc2718**
Sorry in advance - this is kind of a large PR with no super obvious way to split it up into smaller ones. I'll summarize some of the main changes I'm making in this PR below, and if you want to review each of those individually feel free to do that and just approve once you finish them all.

Main changes:

* Tune learning rate for SGD optimizer: see changes to `01_stratified_classification/run_stratified_lasso_penalty.py`, `01_stratified_classification/scripts/run_lasso_lr_compare.sh`, parts of `pancancer_evaluation/prediction/classification.py` and `pancancer_evaluation/utilities/classify_utilities.py`. This turns out to matter quite a bit for SGD performing well, and once we use a slightly more sophisticated approach for tuning the learning rate (constant learning rate + a grid search in this case) we get much better performance, on par with `liblinear`.

* Try a pytorch implementation of SGD: we did this primarily to make sure the SGD performance/regularization dynamics weren't specific to the sklearn implementation. These changes are in `01_stratified_classification/run_stratified_nn.py` and  `pancancer_evaluation/prediction/classification.py` (the `train_mlp_lr` function primarily). We probably won't end up using these results for much in the paper, but it was a useful sanity check.

* Try SGD and liblinear on some simulated data: these are the `01_stratified_classification/sgd_params/sim.ipynb` and `01_stratified_classification/sgd_params/sim_lr.ipynb` notebooks. I used these to iterate quickly on the learning rate changes and compare our results to L2 regularization, but the results turn out to be a bit different on real data so I'm not sure how applicable this simulation approach is to the problem we're trying to address in our paper. I want to keep these scripts around for future reference, though.

I have a summary of the main plots/conclusions in these slides, in case they help a bit with putting the results in context: https://docs.google.com/presentation/d/1LRBq_ciFeS503J8-GeH51l-1p4RTdWJPn_LcGhHNjgM/edit?usp=sharing. Let me know if you have questions!

### CODE REVIEW

File: pancancer_evaluation/utilities/file_utilities.py [hunk 1]
```diff
@@ -9,6 +9,50 @@
...
+                       **kwargs):
+    """Construct a filename from varying experimental parameters.
+
+    The format of the filename will roughly look like this:
+    {output_dir}/{a}_{k}_{file_descriptor}{extension}
+    where a = '_'.join([ar for ar in args]),
+          k = '_'.join(['{k}{v} for k, v in kwargs.items()])
+
+    For example:
+    >>> construct_filename('output_dir', 'output', '.tsv',
```
Comment (arielah) [inline pancancer_evaluation/utilities/file_utilities.py:25]: Is having the ">>>" and "..." for code written inside a docstring standard? It's kind of throwing me off

File: 01_stratified_classification/sgd_params/nbconverted/sim.py [hunk 2]
```diff
@@ -0,0 +1,618 @@
...
+from sklearn.linear_model import LogisticRegression, SGDClassifier
+from sklearn.metrics import roc_auc_score, average_precision_score, log_loss
+from sklearn.model_selection import KFold
+from sklearn.preprocessing import StandardScaler
+
+
+# In[2]:
+
+
+n = 200
```
Comment (arielah) [inline 01_stratified_classification/sgd_params/nbconverted/sim.py:22]: Some explanation of what each of these parameters is would be helpful.

**Review by arielah — APPROVED** [review]
I'm glad you checked this out! Some interesting results in here, and they'll definitely be interesting for the paper.

File: pancancer_evaluation/utilities/file_utilities.py (hunk 1 above)
Comment (jjc2718) [reply pancancer_evaluation/utilities/file_utilities.py:25]: It is - you can actually run them directly as unit tests with the [doctest package](https://docs.python.org/3.4/library/doctest.html) (which is a built-in part of the Python standard library).

File: pancancer_evaluation/prediction/classification.py [hunk 3]
```diff
@@ -163,10 +165,65 @@ def train_lasso(X_train,
...
+                eta0 = 0.005
+            return eta0
+
+        if sgd_lr_schedule == 'constant_search':
+            # in this case, do a grid search to find the best constant learning rate
+            # this only uses the train/valid split, test data isn't used at all here
+            clf_parameters = {
+                "classify__eta0": cfg.constant_search_lr,
+            }
+            clf_pipeline = Pipeline(
```
[the remark this answers is not in the document]
Comment (jjc2718) [reply pancancer_evaluation/prediction/classification.py:199]: Yeah, it's a bit messy to set up for a grid search since you have to nest a classifier inside a grid search object, and in this case I'm then training a separate classifier using the grid search best parameter. It's not ideal but I don't think there are any easy ways to make it less verbose.

File: pancancer_evaluation/prediction/classification.py [hunk 4]
```diff
@@ -163,10 +165,65 @@ def train_lasso(X_train,
...
+        np.arange(X_train.shape[0]),
+        test_size=(1 / n_folds),
+        shuffle=True
+    )
+    X_subtrain, X_valid = X_train.iloc[subtrain_ixs, :], X_train.iloc[valid_ixs, :]
+    y_subtrain, y_valid = y_train.iloc[subtrain_ixs, :], y_train.iloc[valid_ixs, :]
+
     if use_sgd or cfg.lasso_sgd:
+
+        def get_eta0(lr_schedule):
```
[the remark this answers is not in the document]
Comment (jjc2718) [reply pancancer_evaluation/prediction/classification.py:185]: I typically do it when I have a short/simple function that I only need to use in a particular scope, since the nested function won't be usable outside of the function containing it. I mostly just do it for readabillity - in this case I found it a bit syntactically easier to read than just putting the if/else block where I need it.

I do think this code has gotten a bit more complicated than I would like, though. I'll think about ways it could be improved.

File: pancancer_evaluation/prediction/classification.py [hunk 5]
```diff
@@ -188,25 +251,117 @@ def train_lasso(X_train,
...
+                 X_test,
+                 y_train,
+                 y_test,
+                 seed,
+                 hparams={},
+                 n_folds=3,
+                 max_iter=100,
+                 search_n_iter=20):
+
+    import torch.optim
```
[the remark this answers is not in the document]
Comment (jjc2718) [reply pancancer_evaluation/prediction/classification.py:276]: It's mostly that the `torch` package is a bit annoying/complicated to install in a conda environment. If I put the `torch` imports at the top of the file, anyone using my code would have to have it installed and configured, but this way only people that are actually using the neural network functions will need it.

I don't think we're going to include the NN results in the paper, so keeping the imports encapsulated like this makes sense in the context of distributing our code.

File: pancancer_evaluation/utilities/analysis_utilities.py [hunk 6]
```diff
@@ -94,12 +94,19 @@ def load_prediction_results_lasso_range(results_dir,
         gene_dir = os.path.join(results_dir, gene_name)
         if not os.path.isdir(gene_dir): continue
         for results_file in os.listdir(gene_dir):
-            if not ('classify_metrics' in results_file): continue
+            if not ('classify' in results_file): continue
```
[the remark this answers is not in the document]
Comment (jjc2718) [reply pancancer_evaluation/utilities/analysis_utilities.py:97]: Because of the changes I made to how I'm doing file naming in `file_utilities.py`, 'classify' and 'utilities' aren't necessarily always together in the file names. Separating them makes the code work for both the old file naming scheme and the new, more flexible one.
